src/flamenco/txn/fd_txn_generate.c

```diff
@@ -0,0 +1,136 @@
+
+  // Write number of accounts (compact-u16)
+  fd_memcpy( write_ptr, compact_acct_cnt, compact_acct_cnt_sz );
+  write_ptr += compact_acct_cnt_sz;
+
+  // Write accounts list to txn payload
+  ulong signers_write_sz = ACCOUNT_SZ * (accounts->signature_cnt - accounts->readonly_signed_cnt);
+  fd_memcpy( write_ptr, accounts->signers_w, signers_write_sz );
+  write_ptr += signers_write_sz;
+
+  fd_memcpy( write_ptr, accounts->signers_r, ACCOUNT_SZ * accounts->readonly_signed_cnt );
+  write_ptr += ACCOUNT_SZ * accounts->readonly_signed_cnt;
+
+  ulong non_signers_write_sz = ACCOUNT_SZ * (accounts->acct_cnt - accounts->readonly_unsigned_cnt - accounts->signature_cnt);
+  fd_memcpy( write_ptr, accounts->non_signers_w, non_signers_write_sz);
+  write_ptr += non_signers_write_sz;
+
+  fd_memcpy( write_ptr, accounts->non_signers_r, ACCOUNT_SZ * accounts->readonly_unsigned_cnt );
+  write_ptr += ACCOUNT_SZ * accounts->readonly_unsigned_cnt;
+  FD_TEST( (ushort)((ulong)write_ptr - (ulong)out_txn_payload) == txn_meta->recent_blockhash_off );
+
+  // Write recent blockhash
+  if ( opt_recent_blockhash ) {
+    fd_memcpy( write_ptr, opt_recent_blockhash, 32 );
+    write_ptr += 32;
+  }
+
+  return (ulong)(write_ptr - out_txn_payload);
+}
+
+ulong fd_txn_add_instr( uchar * txn_meta_ptr,
+                       uchar out_txn_payload[ static FD_TXN_MTU ],
+                       uchar program_id,
+                       uchar * accounts,
+                       ulong accounts_len,
+                       fd_build_instr_fun instr_fun,
+                       uchar * opt_build_args,
+                       ulong opt_args_len ) {
+
+  fd_txn_t * txn_meta = (fd_txn_t *) txn_meta_ptr;
+  FD_TEST( txn_meta->instr_cnt < FD_TXN_INSTR_MAX );
+  FD_TEST( txn_meta->recent_blockhash_off != 0 );
+
+  uchar * instr_start = out_txn_payload + txn_meta->recent_blockhash_off + BLOCKHASH_SZ;
+  txn_meta->instr_cnt++;
+  uchar compact_instr_cnt[3];
+  uchar compact_instr_cnt_sz = (uchar) fd_cu16_enc( (ushort)txn_meta->instr_cnt, compact_instr_cnt );
+  FD_TEST( compact_instr_cnt_sz == 1 );
+  uchar * write_ptr       = instr_start;
+  fd_memcpy( write_ptr, compact_instr_cnt, compact_instr_cnt_sz );
+  write_ptr += compact_instr_cnt_sz;
+
+  if ( txn_meta->instr_cnt > 1 ) {
+    write_ptr = out_txn_payload + txn_meta->instr[txn_meta->instr_cnt-2].data_off + txn_meta->instr[txn_meta->instr_cnt-2].data_sz;
+  }
+
+  instr_start = write_ptr;
+
+  fd_memcpy( write_ptr, &program_id, sizeof(uchar) );
+  write_ptr += sizeof(uchar);
+
+  uchar compact_accts_len[3];
+  uchar compact_accts_len_sz = (uchar) fd_cu16_enc( (ushort)accounts_len, compact_accts_len );
+  fd_memcpy( write_ptr, compact_accts_len, compact_accts_len_sz );
+  write_ptr += compact_accts_len_sz;
+
+  fd_memcpy( write_ptr, accounts, accounts_len );
+  write_ptr += accounts_len;
+
+  // Build instruction data
+  uchar instr_buf[FD_TXN_MTU];
+  ushort data_sz = (*instr_fun)( instr_buf, opt_build_args, opt_args_len );
```

Comment:
I'm not sure why this callback-style API makes more sense than just taking a pointer. The callback doesn't have any extra information at this point than the caller of this method would have, right? The one potential advantage of the callback API I could see is the ability to serialize directly into the transaction buffer to save a memcpy, but you're not doing that here, so I'm confused.

Reply:
I wrote this to allow for the memcpy to be saved but this was before I knew about the compact u16 stuff :( Can't use the txn buffer directly because we need the size of the instr. Will re-think this

Reply:
Switched to using an instr buf pointer